src/audio/google/google_ctc_audio_processing.c

```diff
@@ -0,0 +1,574 @@
+
+	cd->output_ptr -= samples_to_consume;
+	for (int i = 0; i < cd->output_ptr; ++i)
+		cd->output[i] = cd->output[i + samples_to_consume];
+}
+
+static void ctc_s32_default(struct google_ctc_audio_processing_comp_data *cd,
+			    const struct audio_stream *source,
+			    struct audio_stream *sink,
+			    struct input_stream_buffer *input_buffers,
+			    struct output_stream_buffer *output_buffers,
+			    uint32_t frames)
+{
+	int n_ch = audio_stream_get_channels(source);
+	int samples = frames * n_ch;
+
+	int32_t *src = audio_stream_get_rptr(source);
+	int32_t *dest = audio_stream_get_wptr(sink);
+
+	int samples_to_process = MIN(samples, audio_stream_samples_without_wrap_s32(source, src));
+	int samples_to_consume = 0;
+
+	for (int i = 0; i < samples_to_process; ++i)
+		cd->input[cd->input_ptr++] = convert_int32_to_float(src[i]);
+	src = audio_stream_wrap(source, src + samples_to_process);
+	input_buffers->consumed += audio_stream_frame_bytes(source) * samples_to_process / n_ch;
+
+	while (cd->input_ptr >= cd->num_frames * n_ch) {
+		GoogleCtcAudioProcessingProcess(cd->state, cd->input, cd->output + cd->output_ptr,
+						cd->num_frames, n_ch);
+		cd->output_ptr += cd->num_frames * n_ch;
+
+		cd->input_ptr -= cd->num_frames * n_ch;
+		for (int i = 0; i < cd->input_ptr; ++i)
+			cd->input[i] = cd->input[i + cd->num_frames * n_ch];
+	}
+
+	samples_to_consume = MIN(cd->output_ptr, audio_stream_samples_without_wrap_s32(sink, dest));
+	for (int i = 0; i < samples_to_consume; ++i)
+		dest[i] = convert_float_to_int32(cd->output[i]);
+	dest = audio_stream_wrap(sink, dest + samples_to_consume);
+	output_buffers->size += audio_stream_frame_bytes(sink) * samples_to_consume / n_ch;
+
+	cd->output_ptr -= samples_to_consume;
+	for (int i = 0; i < cd->output_ptr; ++i)
+		cd->output[i] = cd->output[i + samples_to_consume];
+}
+
+/* Processing functions table */
+const struct ctc_func_map ctc_fnmap[] = {
+#if CONFIG_FORMAT_S16LE
+	{ SOF_IPC_FRAME_S16_LE, ctc_s16_default },
+#endif /* CONFIG_FORMAT_S16LE */
+#if CONFIG_FORMAT_S24LE
+	{ SOF_IPC_FRAME_S24_4LE, ctc_s24_default },
+#endif /* CONFIG_FORMAT_S24LE */
+#if CONFIG_FORMAT_S32LE
+	{ SOF_IPC_FRAME_S32_LE, ctc_s32_default },
+#endif /* CONFIG_FORMAT_S32LE */
+};
+
+const size_t ctc_fncount = ARRAY_SIZE(ctc_fnmap);
+
+/**
+ * \brief Retrieves an CTC processing function matching
+ *	  the source buffer's frame format.
+ * \param fmt the frames' format of the source and sink buffers
+ */
+static ctc_func ctc_find_func(enum sof_ipc_frame fmt)
+{
+	int i;
+
+	/* Find suitable processing function from map */
+	for (i = 0; i < ctc_fncount; i++) {
+		if (fmt == ctc_fnmap[i].fmt)
+			return ctc_fnmap[i].func;
+	}
+
+	return NULL;
+}
+
+static int ctc_free(struct processing_module *mod)
+{
+	struct google_ctc_audio_processing_comp_data *cd = module_get_private_data(mod);
+
+	if (cd) {
+		rfree(cd->input);
+		rfree(cd->output);
+		if (cd->state) {
+			GoogleCtcAudioProcessingFree(cd->state);
```

Comment:
Freeing a NULL pointer (which is all GoogleCtcAudioProcessingFree() is a noop, so you can eliminate the if() expression. Also: why set "state" to NULL but not input or output, which are now dangling pointers? 

Normally a "free()" call is a destructor and you assume that no one will touch the struct in the future.  But here you seem to assume that it needs to be re-initialized, which makes me worry about the other fields too.  Basically best to pick one convention and stick to it.

Reply:
Fixed.